README.md
# InordateVM

InordateVM is a minimalistic and complete game-oriented Virtual Machine implementing the Inordate language, with such features as

  - Easy to implement interpreter
  - Basic assembly-like instructions
  - High level game-specific features

# To Do

  - Write C compiler for Inordate
  - Cleanup interpreter folder with obj, src, header and bin folders.
  - Create makefile for interpterer 
  - Combine and sync all readme, todo and changelog docs
  - Compiler: Test to make sure variable numbers do not exceed limit
  - Interpreter: How do I pass parameters, on the stack or with registers? what if i need to preserve registers on the stack>
  - Interpreter: Which math functions should set the flags? Integer only, or floating too?
  - Interpreter: Add pow and div functions
  - Interpreter: Verify cmp, sub, mul, div, test and pow functions
  - Interpreter: Create class and implement physics objects, update on each loop, store in linked list etc.


# Explination

It is worth mentioning here the distinction between the interpreter, the actual virtual machine, and the compiler. While the interpreter takes a file, decodes and executes the instructions, the compiler is responsible for taking Inordate source code and converting it to a binary file.

# Inordate

Inordate is a strongly typed and statically typed assembly-like language. The following is a simple example of Inordate code:

    float $1
    float $2
    float $0
    mul $0, $1, $2
    
which simply declares $0, $1 and $2 as 32-bit floating numbers and stores the result of $2 times $1 into $0. Not only does Inordate contain the basic instructions for completeness, it also contains higher level types and functions for higher level coding. Such examples include vectors, and physics objects which automatically handle moving objects. See [bytecodes.xml](Compiler/src/vmcompiler/bytecodes.xml) for a complete list of Inordate instructions.

# Compiler

The compiler is a Java application designed to convert Inordate source code into bytecode more readable by the interpreter. The compiler is designed to perform type checking, type deduction, and techniques to make coding in Inordate simple. 

# Interpreter 

Since Inordate was desigend to run on simple MCUs or Embedded processors where resources are limited, the interpreter must be light weight. This means that a 'Simple interpreter, Complex compier' approach is taken. 

A time where this philosophy is not adhered to is, for example, the implementation of the physics object, which increases interpreter complexity for coding simplicity.


compiler/changelog.txt
Version 1.0.7
    - Endian-ness of integers in output file changed
    - halt instruction removed from bytecodes.xml, amongst other changes

Version 1.0.6
    - Signature and bytecode hashmap now populated from xml document
    - XML document now integrated as a resource
    - Define directive now works in declarations
    - Define directive implemented on individual arguments, not argument list

Version 1.0.5
    - gitignore updated
    - Project name corrected
    - Javadocs updated
    - Undeclared variables now caught
    - Bytecode values now entered for instructions

Version 1.0.4
    - Preprocessor added
    - Exception handling

Version 1.0.3
    - Compiler now accepts 0 argument instructions without error

Version 1.0.2
    - First parse of source implemented

Version 1.0.1
    - Javadocs added
    - Overhaul of functions
    - Function signatures added via hash map

Version 1.0.0
    - Initial Commit


interpreter/main.cxx
/*
 * main.cxx
 * 
 * Copyright 2019 Will <Will@DESKTOP-PL3AG0U>
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA.
 * 
 * 
 */


#include <iostream> //cout
#include <fstream> //ifstream
#include <stdio.h> //printf
#include <cstdlib> //std::rand
#include <ctime> //std::time
#include <list> //std::list - linked list

#include "functions.h"
#include "physics.h"

const int REGISTER_COUNT = 30;

std::list<PhysicsObject>::iterator iteratorConvert(int num)
{
	return *(std::list<PhysicsObject>::iterator*)&num;
}

int iteratorConvert(std::list<PhysicsObject>::iterator it)
{
	return *(int*)&it;
}

int start_machine(std::ifstream* file, int* registers, Flags* flags, std::list<PhysicsObject>* phys)
{
	using namespace std;
	
	//Get the op code pointed to by the program counter (PC)
	while (true)
	{
		
		
		//Current bytecode
		int fetch;
		
		//Buffer for arguments
		int args[20];
		
		list<PhysicsObject>::iterator it;
		
		//Fetch opcode
		file->read((char*)&fetch, sizeof(fetch));
		
		if (file->eof())
			break;
		
		printf("code %x at %i\n", fetch, (int)file->tellg());
		
		printf("    flags: %i %i %i %i\n", file->good(), file->eof(), file->fail(), file->bad());
		
		
		//Calculate number of arguments for this opcode
		int argcount = (fetch / 256);
		
		cout << "    arg count: " << argcount << endl;
		
		//Populate args with the required number of arguments for opcode
		file->read((char*)args, sizeof(int) * argcount);
		
		cout << "    Args: ";
		for (int i = 0; i < argcount; ++i)
			cout << args[i] << ", ";
		cout << endl;
		
		//Decode and execute instruction
		switch (fetch)
		{
			//Zero argument functions
			case 0x0:
				instr_nop();
				break;
			
			//Single argument functions
			case 0x0100: //jmp_li
				file->seekg(args[0] * sizeof(int)); //Seek to jump location
				break;
			case 0x0101: //je_li
				if (flags->equal)
					file->seekg(args[0] * sizeof(int)); //Seek to jump location
				break;
			case 0x0102: //jne_li
				if (!flags->equal)
					file->seekg(args[0] * sizeof(int)); //Seek to jump location
				break;
			case 0x0103: //jl_li
				if (flags->sign)
					file->seekg(args[0] * sizeof(int)); //Seek to jump location
				break;
			case 0x0104: //jle_li
				if (!flags->sign)
					file->seekg(args[0] * sizeof(int)); //Seek to jump location
				break;
			case 0x0105: //external_li
				
				break;
			case 0x0106: //elapsed_vi
				
				break;
			case 0x0107:
				instr_dec_vi(args, registers, flags);
				break;
			case 0x0108:
				instr_inc_vi(args, registers, flags);
				break;
			case 0x0109:
				instr_random_vi(args, registers);
				break;
			case 0x010A:
				instr_random_vf(args, registers);
				break;
			case 0x010B:
				instr_random_vv(args, registers);
				break;
			case 0x0111:
				instr_test_vi(args, registers, flags);
				break;
			case 0x0112:
				instr_test_vf(args, registers, flags);
				break;
			case 0x0113:
				instr_test_vv(args, registers, flags);
				break;
			case 0x0114:
				it = iteratorConvert(registers[args[0]]);
				
				printf("    %f %f %f\n", it->getMass(), it->getXPos(), it->getYPos());
				break;
				
				
			case 0x0200:
				instr_sin_vf_vf(args, registers);
				break;
			case 0x0201:
				instr_cos_vf_vf(args, registers);
				break;
			case 0x0202:
				instr_tan_vf_vf(args, registers);
				break;
			case 0x0203:
				instr_abs_vi_vi(args, registers);
				break;
			case 0x0204:
				instr_abs_vf_vf(args, registers);
				break;
			case 0x0205:
				instr_norm_vf_vf(args, registers);
				break;
			case 0x0206:
				instr_norm_vf_vf(args, registers);
				break;
			case 0x0207:
				instr_ln_vf_vf(args, registers);
				break;
			case 0x0208:
				instr_exp_vf_vf(args, registers);
				break;
			case 0x0209:
				instr_floor_vf_vf(args, registers);
				break;
			case 0x020A:
				instr_ceil_vf_vf(args, registers);
				break;
			case 0x020B:
				instr_sqrt_vf_vf(args, registers);
				break;
			case 0x020C:
				instr_neg_vi_vi(args, registers);
				break;
			case 0x020D:
				instr_neg_vf_vf(args, registers);
				break;
				
				
			case 0x0240:
				instr_store_vi_li(args, registers);
				break;
			case 0x0241:
				instr_store_vf_lf(args, registers);
				break;
			case 0x0242:
				instr_store_vi_vi(args, registers);
				break;
			case 0x0243:
				instr_store_vi_vf(args, registers);
				break;
			case 0x0244:
				instr_store_vf_vi(args, registers);
				break;
			case 0x0245:
				instr_store_vf_vf(args, registers);
				break;
				
			case 0x0246:
				instr_cmp_vi_li(args, registers, flags);
				break;
			case 0x0247:
				instr_cmp_vf_lf(args, registers, flags);
				break;
			case 0x0248:
				instr_cmp_vi_vi(args, registers, flags);
				break;
			case 0x0249:
				instr_cmp_vi_vf(args, registers, flags);
				break;
			case 0x024A:
				instr_cmp_vf_vi(args, registers, flags);
				break;
			case 0x024B:
				instr_cmp_vf_vf(args, registers, flags);
				break;
			
				
			case 0x0300:
				instr_add_vi_vi_vi(args, registers, flags);
				break;
			case 0x0301:
				instr_add_vf_vf_vf(args, registers, flags);
				break;
			case 0x0302:
				instr_add_vv_vv_vv(args, registers, flags);
				break;
			case 0x0303:
				instr_add_vi_vi_li(args, registers, flags);
				break;
			case 0x0304:
				instr_add_vf_vf_lf(args, registers, flags);
				break;
				
			case 0x0305:
				instr_sub_vi_vi_vi(args, registers, flags);
				break;
			case 0x0306:
				instr_sub_vf_vf_vf(args, registers, flags);
				break;
			case 0x0307:
				instr_sub_vv_vv_vv(args, registers, flags);
				break;
			case 0x0308:
				instr_sub_vi_vi_li(args, registers, flags);
				break;
			case 0x0309:
				instr_sub_vf_vf_lf(args, registers, flags);
				break;
			case 0x030A:
				instr_sub_vi_li_vi(args, registers, flags);
				break;
			case 0x030B:
				instr_sub_vf_lf_vf(args, registers, flags);
				break;
				
			case 0x030C:
				instr_mul_vi_vi_vi(args, registers, flags);
				break;
			case 0x030D:
				instr_mul_vf_vf_vf(args, registers, flags);
				break;
			case 0x030E:
				instr_mul_vv_vv_vf(args, registers, flags);
				break;
			case 0x030F:
				instr_mul_vv_vv_lf(args, registers, flags);
				break;
				
			case 0x0310:
				instr_mul_vi_vi_li(args, registers, flags);
				break;
			case 0x0311:
				instr_mul_vf_vf_lf(args, registers, flags);
				break;
				
				
				
			case 0x0400:
				registers[args[0]] = iteratorConvert(phys->insert(phys->end(), PhysicsObject(1.0f, Vector2f(0.4f, 0.0f))));
				break;
				
			default:
				printf("Error: Bytecode 0x%08X not recognised. Aborting VM\n", fetch);
				return 1;
				
		}
		
		cout << "    registers: ";
		
		for (int i = 0; i < 4; ++i)
			printf("%i ", registers[i]);
			
		printf("%g\n", hardConvert(registers[4]));
			
	}
	
	//Clear eof and error flags
	file->clear();
	
	//Reset stream 
	file->seekg(0);
	
	return 0;
}

void validator()
{
	int args[20];
	int reg[30];
	Flags flags(false, false);
	
	//Use resister 0 for first arg
	args[0] = 0;
	args[1] = 1;
	
	reg[1] = hardConvert(1.33f);
	reg[2] = 0;
	
	instr_store_vf_vf(args, reg);
	
	printf("Integer: %i\n", reg[0]);
	printf("Float  : %g\n", hardConvert(reg[0]));
	
}


int main(int argc, char **argv)
{
	using namespace std;
	
	char* setup_path = argv[1];
	char* loop_path = argv[2];
	
	int registers[30];
	
	list<PhysicsObject> phys_obj;
	
	ifstream setup(setup_path, ifstream::binary);
	ifstream loop(loop_path, ifstream::binary);
	
	srand(time(nullptr));
	
	Flags flags(false, false);
	
	//Get the file path from command line args
	if (argc != 3)
	{
		cout << "Error: Invalid number of command line arguments. Aborting VM." << endl;
		return 1;
	}
	
	cout << "START SETUP ROUTINE" << endl;
	if (start_machine(&setup, registers, &flags, &phys_obj))
		return 1;
		
	cout << "START LOOP ROUTING" << endl;
	
	for (int i = 0; i < 10; ++i)
	{
		if (start_machine(&loop, registers, &flags, &phys_obj))
			return 1;
	}
	
	
	
	return 0;
}



interpreter/include/physics.h

#include <limits>
#include "vmath.h"

#include <iostream>

class PhysicsObject
{
private:
	Vector2f _acceleration;
	Vector2f _force;
	Vector2f _velocity;
	Vector2f _position;
	
	float _topSpeed;
	float _mass;
	float _friction;
	
public:
	PhysicsObject(float mass, Vector2f pos, Vector2f force = Vector2f(0.0f, 0.0f), Vector2f vel = Vector2f(0.0f, 0.0f), float max = std::numeric_limits<double>::infinity(), float friction = 0.0f)
	{
		_mass = mass;
		_force = force;
		_acceleration = _force / _mass;
		_velocity = vel;
		_position = pos;
		
		_topSpeed = max;
		_friction = friction;
		
	}
	
	void setForce(const Vector2f &force)
	{
		_force = force;
	}
	
	void setVelocity(const Vector2f &vel)
	{
		_velocity = vel;
	}
	
	void addForce(const Vector2f &force)
	{
		_force += force;
	}
	
	void setFriction(float friction)
	{
		_friction = friction;
	}
	
	void update(int elapsed)
	{
		using namespace std;
		
		_acceleration = _force / _mass;
		_velocity += _acceleration * elapsed / 1000.0f;
		_position += _velocity * elapsed / 1000.0f;
		
		cout << "    Mass: " << _mass << ". Acceleration: " << _acceleration.x << ", " << _acceleration.y << ". Velocity: "
			 << _velocity.x << ". " << _velocity.y << ". Position: "<< _position.x << ", " << _position.y << endl;
	}
	
	float getMass() const { return _mass; }
	float getXPos() const { return _position.x; }
	float getYPos() const { return _position.y; }
	
	
};


interpreter/src/main.cpp
/*
 * main.cxx
 * 
 * Copyright 2019 Will <Will@DESKTOP-PL3AG0U>
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA.
 * 
 * 
 */


#include <iostream> //cout
#include <fstream> //ifstream
#include <stdio.h> //printf
#include <cstdlib> //std::rand
#include <ctime> //std::time
#include <list> //std::list - linked list

#include "functions.h"
#include "physics.h"

const int REGISTER_COUNT = 30;

const int MAX_ARG_COUNT = 20;

std::list<PhysicsObject>::iterator iteratorConvert(int num)
{
	return *(std::list<PhysicsObject>::iterator*)&num;
}

int iteratorConvert(std::list<PhysicsObject>::iterator it)
{
	return *(int*)&it;
}

int start_machine(std::ifstream* file, int* registers, Flags* flags, std::list<PhysicsObject>* phys)
{
	using namespace std;
	
	//Get the op code pointed to by the program counter (PC)
	while (true)
	{
		//Current bytecode
		int fetch;
		
		//Buffer for arguments
		int args[MAX_ARG_COUNT];
		
		
		
		//Fetch opcode
		file->read((char*)&fetch, sizeof(fetch));
		
		if (file->eof())
			break;
		
		printf("code %x at %i\n", fetch, (int)file->tellg());
		
		
		//Calculate number of arguments for this opcode
		int argcount = (fetch / 256);
		
		cout << "    arg count: " << argcount << endl;
		
		//Populate args with the required number of arguments for opcode
		file->read((char*)args, sizeof(int) * argcount);
		
		cout << "    Args: ";
		for (int i = 0; i < argcount; ++i)
			cout << args[i] << ", ";
		cout << endl;
		
		//Decode and execute instruction
		switch (fetch)
		{
			//Zero argument functions
			case 0x0:
				instr_nop();
				break;
			
			//Single argument functions
			case 0x0100: //jmp_li
				file->seekg(args[0] * sizeof(int)); //Seek to jump location
				break;
			case 0x0101: //je_li
				if (flags->equal)
					file->seekg(args[0] * sizeof(int)); //Seek to jump location
				break;
			case 0x0102: //jne_li
				if (!flags->equal)
					file->seekg(args[0] * sizeof(int)); //Seek to jump location
				break;
			case 0x0103: //jl_li
				if (flags->sign)
					file->seekg(args[0] * sizeof(int)); //Seek to jump location
				break;
			case 0x0104: //jle_li
				if (!flags->sign)
					file->seekg(args[0] * sizeof(int)); //Seek to jump location
				break;
			case 0x0105: //external_li
				
				break;
			case 0x0106: //elapsed_vi
				
				break;
			case 0x0107:
				instr_dec_vi(args, registers, flags);
				break;
			case 0x0108:
				instr_inc_vi(args, registers, flags);
				break;
			case 0x0109:
				instr_random_vi(args, registers);
				break;
			case 0x010A:
				instr_random_vf(args, registers);
				break;
			case 0x010B:
				instr_random_vv(args, registers);
				break;
			case 0x0111:
				instr_test_vi(args, registers, flags);
				break;
			case 0x0112:
				instr_test_vf(args, registers, flags);
				break;
			case 0x0113:
				instr_test_vv(args, registers, flags);
				break;
			case 0x0114:
				{list<PhysicsObject>::iterator it = iteratorConvert(registers[args[0]]);
				phys->erase(it);}
				break;
				
				
			case 0x0200:
				instr_sin_vf_vf(args, registers);
				break;
			case 0x0201:
				instr_cos_vf_vf(args, registers);
				break;
			case 0x0202:
				instr_tan_vf_vf(args, registers);
				break;
			case 0x0203:
				instr_abs_vi_vi(args, registers);
				break;
			case 0x0204:
				instr_abs_vf_vf(args, registers);
				break;
			case 0x0205:
				instr_norm_vf_vf(args, registers);
				break;
			case 0x0206:
				instr_norm_vf_vf(args, registers);
				break;
			case 0x0207:
				instr_ln_vf_vf(args, registers);
				break;
			case 0x0208:
				instr_exp_vf_vf(args, registers);
				break;
			case 0x0209:
				instr_floor_vf_vf(args, registers);
				break;
			case 0x020A:
				instr_ceil_vf_vf(args, registers);
				break;
			case 0x020B:
				instr_sqrt_vf_vf(args, registers);
				break;
			case 0x020C:
				instr_neg_vi_vi(args, registers);
				break;
			case 0x020D:
				instr_neg_vf_vf(args, registers);
				break;
				
				
			case 0x0240:
				instr_store_vi_li(args, registers);
				break;
			case 0x0241:
				instr_store_vf_lf(args, registers);
				break;
			case 0x0242:
				instr_store_vi_vi(args, registers);
				break;
			case 0x0243:
				instr_store_vi_vf(args, registers);
				break;
			case 0x0244:
				instr_store_vf_vi(args, registers);
				break;
			case 0x0245:
				instr_store_vf_vf(args, registers);
				break;
				
			case 0x0246:
				instr_cmp_vi_li(args, registers, flags);
				break;
			case 0x0247:
				instr_cmp_vf_lf(args, registers, flags);
				break;
			case 0x0248:
				instr_cmp_vi_vi(args, registers, flags);
				break;
			case 0x0249:
				instr_cmp_vi_vf(args, registers, flags);
				break;
			case 0x024A:
				instr_cmp_vf_vi(args, registers, flags);
				break;
			case 0x024B:
				instr_cmp_vf_vf(args, registers, flags);
				break;
			
				
			case 0x0300:
				instr_add_vi_vi_vi(args, registers, flags);
				break;
			case 0x0301:
				instr_add_vf_vf_vf(args, registers, flags);
				break;
			case 0x0302:
				instr_add_vv_vv_vv(args, registers, flags);
				break;
			case 0x0303:
				instr_add_vi_vi_li(args, registers, flags);
				break;
			case 0x0304:
				instr_add_vf_vf_lf(args, registers, flags);
				break;
				
			case 0x0305:
				instr_sub_vi_vi_vi(args, registers, flags);
				break;
			case 0x0306:
				instr_sub_vf_vf_vf(args, registers, flags);
				break;
			case 0x0307:
				instr_sub_vv_vv_vv(args, registers, flags);
				break;
			case 0x0308:
				instr_sub_vi_vi_li(args, registers, flags);
				break;
			case 0x0309:
				instr_sub_vf_vf_lf(args, registers, flags);
				break;
			case 0x030A:
				instr_sub_vi_li_vi(args, registers, flags);
				break;
			case 0x030B:
				instr_sub_vf_lf_vf(args, registers, flags);
				break;
				
			case 0x030C:
				instr_mul_vi_vi_vi(args, registers, flags);
				break;
			case 0x030D:
				instr_mul_vf_vf_vf(args, registers, flags);
				break;
			case 0x030E:
				instr_mul_vv_vv_vf(args, registers, flags);
				break;
			case 0x030F:
				instr_mul_vv_vv_lf(args, registers, flags);
				break;
				
			case 0x0310:
				instr_mul_vi_vi_li(args, registers, flags);
				break;
			case 0x0311:
				instr_mul_vf_vf_lf(args, registers, flags);
				break;
				
			case 0x0340:
			
				break;
				
			case 0x0341:
				{list<PhysicsObject>::iterator it = iteratorConvert(registers[args[0]]);
				it->setForce(Vector2f(hardConvert(registers[args[1]]), hardConvert(registers[args[2]])));}
				break;
			case 0x0342:
				{list<PhysicsObject>::iterator it = iteratorConvert(registers[args[0]]);
				it->setForce(Vector2f(hardConvert(args[1]), hardConvert(args[2])));}
				break;
			case 0x0343:
				{list<PhysicsObject>::iterator it = iteratorConvert(registers[args[0]]);
				it->setVelocity(Vector2f(hardConvert(registers[args[1]]), hardConvert(registers[args[2]])));}
				break;
			case 0x0344:
				{list<PhysicsObject>::iterator it = iteratorConvert(registers[args[0]]);
				it->setVelocity(Vector2f(hardConvert(args[1]), hardConvert(args[2])));}
				break; 
				
				
			case 0x0400:
				registers[args[0]] = iteratorConvert(phys->insert(phys->end(), 
					PhysicsObject(hardConvert(registers[args[1]]), 
						Vector2f(hardConvert(registers[args[2]]), hardConvert(registers[args[3]])))));
				break;
			case 0x0401:
				registers[args[0]] = iteratorConvert(phys->insert(phys->end(), 
					PhysicsObject(hardConvert(args[1]), 
						Vector2f(hardConvert(args[2]), hardConvert(args[3])))));
				break;
				
			default:
				printf("Error: Bytecode 0x%08X not recognised. Aborting VM\n", fetch);
				return 1;
				
		}
		
		cout << "    registers: ";
		
		for (int i = 0; i < 6; ++i)
			printf("%i (%g) ", registers[i], hardConvert(registers[i]));
			
		printf("%g\n", hardConvert(registers[6]));
			
	}
	
	//Clear eof and error flags
	file->clear();
	
	//Reset stream 
	file->seekg(0);
	
	return 0;
}

void validator()
{
	int args[20];
	int reg[30];
	Flags flags(false, false);
	
	//Use resister 0 for first arg
	args[0] = 0;
	args[1] = 1;
	
	reg[1] = hardConvert(1.33f);
	reg[2] = 0;
	
	instr_store_vf_vf(args, reg);
	
	printf("Integer: %i\n", reg[0]);
	printf("Float  : %g\n", hardConvert(reg[0]));
	
}


int main(int argc, char **argv)
{
	using namespace std;
	
	int registers[REGISTER_COUNT];
	
	list<PhysicsObject> phys_obj;
	
	ifstream setup(argv[1], ifstream::binary); //Get setup file
	ifstream loop(argv[2], ifstream::binary); //Get loop file
	
	srand(time(nullptr));
	
	Flags flags(false, false);
	
	//Get the file path from command line args
	if (argc != 3)
	{
		cout << "Error: Invalid number of command line arguments. Aborting VM." << endl;
		return 1;
	}
	
	cout << "START SETUP ROUTINE" << endl;
	if (start_machine(&setup, registers, &flags, &phys_obj))
		return 1;
		
	cout << "START LOOP ROUTING" << endl;
	
	for (int i = 0; i < 10; ++i)
	{
		cout << "Physics Updates" << endl;
		for (list<PhysicsObject>::iterator i = phys_obj.begin(); i != phys_obj.end(); i++)
			i->update(1000);
		
		if (start_machine(&loop, registers, &flags, &phys_obj))
			return 1;
	}
	
	
	return 0;
}

